Fix to correctly locate python DLL on MS-Windows

ikappaki opened the pull request:
Hi,

can you pleaser review patch to properly locate the official python dll library on MS-Windows. It fixes #246. 

The python dll name on MS-Windows does not have `.` in its name as it is the case with the other platforms. This just adds a version of the library to look out for without the `.`.

In the meantime, I've noticed that there are no cross platform CI tests on the codebase (travis config appears to be outdated and limited ubuntu only), and though to add an example github workflow to run the unit tests across macos, win and Linux. I've only included the latest official python version (3.11) for economy, while extending the test coverage to at least 3 jdk versions (8, 17 and 19) which I thought as being important coverage. Let me know what you think, happy to revert or further modify. 

I've also updated one of the test to import `Mappings` from `collections` to `collections.abc` as it fails in python versions > 3.9 (see https://stackoverflow.com/questions/70195545/module-collections-has-no-attribute-mapping-issue-on-macos-for-sdk-installa for reference)

Thanks

jjtolton commented on .gitignore:
```diff
@@ -20,4 +20,5 @@ a.out
-pregen-ffi-test
\ No newline at end of file
+pregen-ffi-test
+*~
```
This change seems unrelated to the patch

ikappaki replied:
Yes, sorry I forgot to mention this. 

files ending in `~` are files created by the Emacs editors as backup of the files being edited. They are typically ignored otherwise they appear all over the place in git status. Since a great many Clojure developers are using Emacs as their IDE of choice, I thought it was safe to ignore them. I can revert if you think this is unnecessary or too narrow a view.

jjtolton replied:
I'm also an emacs user.  While I appreciate the sentiment, let's keep the revisions related to your primary change just to keep the history as clean as possible.  A PR with this addition would be welcome. (also note that you can [disable the creation of those tilde files in emacs](https://stackoverflow.com/questions/2680389/how-to-remove-all-files-ending-with-made-by-emacs#2680682). Sorry to nitpick but since this is a public repo want to make an effort to the users to keep the history clean.

ikappaki replied:
Change reverted.

jjtolton commented on test/libpython_clj2/python_test.clj:
```diff
@@ -365,7 +365,7 @@ class Foo:
         bridged-iter (py/as-python (repeat 5 1))
         bridged-list (py/as-python (vec (range 10)))
-        pycol (py/import-module "collections")
+        pycol (py/import-module "collections.abc")
```
Unless I'm missing something, this looks like it may have been left in by accident.  Is there a reason for changing to `collections.abc` that is related to this patch?

ikappaki replied:
Actually, I have mentioned this in the opening comment, but it seems I did a bad job at explaining it.

The `pycol` above is used to import `Mapping` (in the next line) from the `collections` package, instead of the `collections.abc` package that it belongs to.

This does not work any more with python 3.10 and above

https://github.com/python/cpython/blob/38b489bae817eef648fa14e775296c6245da2cb2/Lib/collections/__init__.py#L55

``` python
warnings.warn("Using or importing the ABCs from 'collections' instead "
                      "of from 'collections.abc' is deprecated since Python 3.3, "
                      "and in 3.10 it will stop working",
                      DeprecationWarning, stacklevel=2)
```


As such,  this particular `libpython-clj`  fail when invoked with pytnon 3.10+.  Changing the import to `collections.abc` makes it work with any python version.

jjtolton replied:
While not related to the purpose of the PR, it is a great catch.  If you put it in a separate PR, I would be happy to sign off.  I don't mean to be discouraging, all three of these changes are great.  But as it's a public repo we owe it to the users to keep the history as clean as possible.

ikappaki replied:
Change reverted and opened #248

jjtolton commented:
Over looks good!  Thank you @ikappaki.  @cnuernber could you comment on the inclusion of [this file](https://github.com/clj-python/libpython-clj/blob/56673740767f3274335519177d54fe692b055691/.github/workflows/test.yml)? Not sure how it fits in with the overall repo management.

jjtolton reviewed (changes requested):
Overall looks good!  Could you please either revert or explain those two changes in the `.gitignore` file and the `python_test.clj` file?

ikappaki commented:
> Overall looks good! Could you please either revert or explain those two changes in the `.gitignore` file and the `python_test.clj` file?

Hi @jjtolton , thanks for the thorough review. I've tried explaining these two in their respective comments, I can revert `.gitignore` at your discretion, but `python_test.clj` change is essential for running the tests with python 3.10 and above.

ikappaki commented:
Hi @jjtolton, just to confirm the plan in terms of Issues and PRs to raise and submit going forward as per your suggestion

1. An Issue and PR for the DLL issue (this one)
2. An Issue and PR for the test fix
3. An Issue and PR for the CI coverage upgrade
4. An Issue and PR for the Emacs backup file exclusion

thanks

ikappaki commented:
> Perfect. I'd be comfortable bundling the test job and the test fix into one (items 2 and 3).

I've reverted the test fix, CI suggestion and .gitignore change, and moved the first two in #249.

Thanks

jjtolton commented:
Thank you for this, @ikappaki !
